### PR TITLE
Optimize artifact lookup in DefaultProjectDependencyAnalyzer

### PR DESCRIPTION
**Pull request by sanjana2505006**
While reviewing the core analysis logic, I noticed a performance bottleneck in `DefaultProjectDependencyAnalyzer`. \n\nPreviously, it used a linear search to find which artifact a class belongs to ($O(N)$ for every class usage), which could be a bottleneck in large projects with many dependencies. These changes switch to a `Map`-based lookup ($O(1)$) and modernize the code using `computeIfAbsent` and more robust scope checks.\n\nThis significantly improves the efficiency of the analysis process for projects with large dependency trees.

### CODE REVIEW

File: src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java [hunk 1]
```diff
@@ -260,13 +259,16 @@ private static boolean includedInJDK(Artifact artifact) {
...
-    private static Artifact findArtifactForClassName(Map<Artifact, Set<String>> artifactClassMap, String className) {
+    private static Map<String, Artifact> buildClassToArtifactMap(Map<Artifact, Set<String>> artifactClassMap) {
+        Map<String, Artifact> classToArtifactMap = new HashMap<>();
+
         for (Map.Entry<Artifact, Set<String>> entry : artifactClassMap.entrySet()) {
-            if (entry.getValue().contains(className)) {
-                return entry.getKey();
+            Artifact artifact = entry.getKey();
+            for (String className : entry.getValue()) {
+                classToArtifactMap.put(className, artifact);
```
[the remark this answers is not in the document]
Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java]: @sanjana2505006 please look for this one

[the remark this answers is not in the document]
Comment (sanjana2505006) [reply src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java]: I've updated it to `putIfAbsent` in the latest commit to preserve the original first-match-wins semantics. Thank you  for flagging this, @slawekjaranowski!

File: src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java [hunk 2]
```diff
@@ -226,19 +230,14 @@ private static Set<Artifact> buildDeclaredArtifacts(MavenProject project) {
...
             if (artifact != null && !includedInJDK(artifact)) {
-                Set<DependencyUsage> classesFromArtifact = usedArtifacts.get(artifact);
-                if (classesFromArtifact == null) {
-                    classesFromArtifact = new HashSet<>();
-                    usedArtifacts.put(artifact, classesFromArtifact);
-                }
-                classesFromArtifact.add(classUsage);
+                usedArtifacts.computeIfAbsent(artifact, k -> new HashSet<>()).add(classUsage);
             }
         }
```
[the remark this answers is not in the document]
Comment (sanjana2505006) [reply src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java]: That's a fair point. I'll add focused unit tests covering the new class-to-artifact resolution path, including duplicate-class scenarios, to ensure the chosen artifact is deterministic. Will update it.

**Comment by sanjana2505006** [comment]
Whenever you get a free moment, I would love to hear your thoughts on this. Let me know if anything needs to be updated!

File: src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (slachiewicz) [reply src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java]: still no unit test

[the remark this answers is not in the document]
Comment (sanjana2505006) [reply src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java]: I have just added comprehensive unit tests in `DefaultProjectDependencyAnalyzerTest.java` to cover the new class-to-artifact resolution path.

File: src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java [hunk 3]
```diff
@@ -43,7 +43,9 @@
 import org.apache.maven.project.MavenProject;
 
 /**
- * <p>DefaultProjectDependencyAnalyzer class.</p>
+ * <p>
+ * DefaultProjectDependencyAnalyzer class.
+ * </p>
```
[the remark this answers is not in the document]
Comment (sanjana2505006) [reply src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java]: I've reverted the class Javadoc to use the single-line

**Comment by slawekjaranowski** [comment]
@sanjana2505006 pleas check build output

**Comment by sanjana2505006** [comment]
> @sanjana2505006 pleas check build output

I have an exam shortly, but I'll make sure to look into the build output and resolve this as a priority once I'm done.

**Comment by slawekjaranowski** [comment]
@sanjana2505006 rebased, and fix format

**Comment by sanjana2505006** [comment]
@slawekjaranowski Thank you for the help with the rebase and spotless fix! I’m tied up with my last exam until the 15th, but I'll check back in as soon as I'm done.

**Comment by slawekjaranowski** [comment]
> @slawekjaranowski Thank you for the help with the rebase and spotless fix! I’m tied up with my last exam until the 15th, but I'll check back in as soon as I'm done.

don't worry, take care about your exam

**Review by slawekjaranowski — APPROVED** [review]
